blogs/optimizing-gen-ai-apps-with-durable-semantic-cache/README.md

```diff
@@ -0,0 +1,159 @@
+# Persistent semantic cache example application
+
+This solution automates the provisioning of an application that uses persistent semantic cache powered by Amazon MemoryDB. The application implements an API service to answer questions related to the AWS well architected framework. Once a question has been answered, it is saved to the persistent semantic cache, improving latency and compute cost for the subsequent semantically similar questions.
+
+## Solution components
+
+* Amazon MemoryDB: OSS Redis compatible, durable, in-memory database service that delivers ultra-fast performance. 
+* Amazon API gateway: a fully managed service that makes it easy for developers to create, publish, maintain, monitor, and secure APIs at any scale
+* Amazon Lambda: a serverless compute service for running code without having to provision or manage servers.
+* Amazon Bedrock: a fully managed service that offers a choice of high-performing foundation models (FMs) from leading AI companies
+* Amazon S3: an object storage service offering industry-leading scalability, data availability, security, and performance. Amazon S3 will be used to store the documents that will be ingested into Knowledge Bases for Amazon Bedrock
+* Amazon OpenSearch Serverless: run petabyte-scale workloads without configuring, managing, and scaling OpenSearch clusters. 
+* Knowledge Bases for Amazon Bedrock: provides foundation models contextual information from data sources, such as OpenSearch clusters, using Retrieval-Augmented Generation (RAG)
+
+## Pre-requisites  
+
+* AWS Command Line Interface (AWS CLI)
+* Terraform
+* Amazon Bedrock model access enabled for Anthropic Claude 3 Sonnet and Amazon Titan Embeddings G1 in us-east-1
+* jq and git
+* curl and time commands (installed by default in Mac and Linux systems)
+* Python3.11 (Runtime can be changed to other python versions in the lambda module definition in the main.tf file)
+
+## Installation
+
+1. Clone repo
+```
+git clone https://github.com/aws-samples/amazon-memorydb-for-redis-samples.git
+```
+
+2. Init terraform and apply plan. This will create the infrastructure for the components of the solution
+
+```
+cd amazon-memorydb-for-redis-samples/blogs/optimizing-gen-ai-apps-with-durable-semantic-cache/terraform
+terraform init
+terraform apply --auto-approve
+```
+
+3. Setup knowledge base
+
+The knowledge base is created as part of the terraform plan, but files need to be uploaded to S3 and the ingestion job started
+
+Get the terraform outputs
+```
+export DATA_SOURCE_ID=$(terraform output -raw data_source_id | cut -d, -f1)
+echo "DATA_SOURCE_ID: $DATA_SOURCE_ID"
+export KNOWLEDGE_BASE_ID=$(terraform output -raw knowledge_base_id)
+echo "KNOWLEDGE_BASE_ID: $KNOWLEDGE_BASE_ID"
+export S3_BUCKET_NAME=$(terraform output -raw s3_bucket)
+echo "Bucket: $S3_BUCKET_NAME"
+
+```
+
+Upload docs and start the ingestion job
```

Comment:
I would add in here that this might fail without the proper role / or assume credentials for running the below job. 

#aws sts get-caller-identity